app/src/main/java/org/thoughtcrime/securesms/database/GroupDatabase.java

```diff
@@ -0,0 +1,1372 @@
+   */
+  public void fixMissingMasterKey(@NonNull GroupMasterKey groupMasterKey) {
+    GroupId.V2 groupId = GroupId.v2(groupMasterKey);
+
+    if (getGroupV1ByExpectedV2(groupId).isPresent()) {
+      Log.w(TAG, "There already exists a V1 group that should be migrated into this group. But if the recipient already exists, there's not much we can do here.");
+    }
+
+    SQLiteDatabase db = databaseHelper.getSignalWritableDatabase();
+
+    db.beginTransaction();
+    try {
+      String        query  = GROUP_ID + " = ?";
+      String[]      args   = SqlUtil.buildArgs(groupId);
+      ContentValues values = new ContentValues();
+
+      values.put(V2_MASTER_KEY, groupMasterKey.serialize());
+
+      int updated = db.update(TABLE_NAME, values, query, args);
+
+      if (updated < 1) {
+        Log.w(TAG, "No group entry. Creating restore placeholder for " + groupId);
+        create(groupMasterKey,
+               DecryptedGroup.newBuilder()
+                             .setRevision(GroupsV2StateProcessor.RESTORE_PLACEHOLDER_REVISION)
+                             .build(),
+               true);
+      } else {
+        Log.w(TAG, "Had a group entry, but it was missing a master key. Updated.");
+      }
+
+      db.setTransactionSuccessful();
+    } finally {
+      db.endTransaction();
+    }
+
+    Log.w(TAG, "Scheduling request for latest group info for " + groupId);
+    ApplicationDependencies.getJobManager().add(new RequestGroupV2InfoJob(groupId));
+  }
+
+  /**
+   * @param groupMasterKey null for V1, must be non-null for V2 (presence dictates group version).
+   */
+  private void create(@NonNull GroupId groupId,
+                      @Nullable String title,
+                      @NonNull Collection<RecipientId> memberCollection,
+                      @Nullable SignalServiceAttachmentPointer avatar,
+                      @Nullable String relay,
+                      @Nullable GroupMasterKey groupMasterKey,
+                      @Nullable DecryptedGroup groupState)
+  {
+    RecipientDatabase recipientDatabase = DatabaseFactory.getRecipientDatabase(context);
+    RecipientId       groupRecipientId  = recipientDatabase.getOrInsertFromGroupId(groupId);
+    List<RecipientId> members           = new ArrayList<>(new HashSet<>(memberCollection));
+
+    Collections.sort(members);
+
+    ContentValues contentValues = new ContentValues();
+    contentValues.put(RECIPIENT_ID, groupRecipientId.serialize());
+    contentValues.put(GROUP_ID, groupId.toString());
+    contentValues.put(TITLE, title);
+    contentValues.put(MEMBERS, RecipientId.toSerializedList(members));
+
+    if (avatar != null) {
+      contentValues.put(AVATAR_ID, avatar.getRemoteId().getV2().get());
+      contentValues.put(AVATAR_KEY, avatar.getKey());
+      contentValues.put(AVATAR_CONTENT_TYPE, avatar.getContentType());
+      contentValues.put(AVATAR_DIGEST, avatar.getDigest().orNull());
+    } else {
+      contentValues.put(AVATAR_ID, 0);
+    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 555:562, 618:625.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.